src/xtgeo/grid3d/_gridprop_import_grdecl.py

```diff
@@ -83,6 +80,97 @@ def import_bgrdecl_prop(self, pfile, name="unknown", grid=None):
     pfile.cfclose()
 
 
+@contextmanager
+def open_grdecl(grdecl_file, keywords):
+    """
+    Opens the grdecl_file for reading. Is a generator for tuples of keyword /
+    values in records of that file. The format of the file must be that of the
+    GRID section of a eclipse input DATA file.
+
+    The records looked for must be "simple" ie.  start with the keyword, be
+    followed by single word values and ended by a slash ('/').
+
+    open_grdecl does not follow includes, obey skips, parse MESSAGE commands or
+    make exception for groups and subrecords.
+
+    Raises:
+        ValueError: when end of file is reached without terminating a keyword.
+
+    Args:
+        keywords (List[str]): Which keywords to look for, these are expected
+            to be at the start of a line in the file and be followed by
+        grdecl_file (str or pathlib.Path): a File in grdecl format.
+
+    """
+
+    def read_grdecl(grdecl_stream):
+        words = []
+        keyword = None
+
+        line = grdecl_stream.readline()
+        while line:
+            if keyword is None:
+                splitted = line.split()
+                if splitted:
+                    first_word = splitted[0]
+                    if first_word in keywords:
+                        keyword = first_word
+            else:
+                for word in line.split():
+                    if word == "--":
+                        break
+
+                    if word == "/":
+                        yield (keyword, words)
+                        words = []
+                        keyword = None
+                        break
+                    words.append(word)
+            line = grdecl_stream.readline()
+
+        if keyword is not None:
+            raise ValueError(f"Reached end of stream while reading {keyword}")
+
+    with open(grdecl_file, "r") as stream:
+        yield read_grdecl(stream)
+
+
+def read_grdecl_3d_property(filename, keyword, dimensions, dtype=float):
+    """
+    Read a 3d grid property from a grdecl file, see open_grdecl for description
+    of format.
+
+    Args:
+        filename (pathlib.Path or str): File in grdecl format.
+        keyword (str): The keyword of the property in the file
+        dimensions ((int,int,int)): Triple of the size of grid.
+        dtype (function): The datatype to be read, ie., float.
+
+    Raises:
+        xtgeo.KeywordNotFoundError: If keyword is not found in the file.
+
+    Returns:
+        numpy array with given dimensions and data type read
+        from the grdecl file.
+    """
+    result = None
+
+    with open_grdecl(filename, keywords=(keyword,)) as kw_generator:
+        for key, words in kw_generator:
+            if key == keyword:
```

Comment:
As we are passing in the keyword to be used, can this ever be false?

Reply:
Indeed, except in the missing keyword case. Will fix

Reply:
Done.

Reply:
Not sure if my proposal was the best (my intention was just to make sure I hadn't misunderstood it), feel free to write it how you want for best readability :)